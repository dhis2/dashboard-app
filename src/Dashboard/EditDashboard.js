import React, { Fragment, Component } from 'react';
import { connect } from 'react-redux';
import { fromDashboards, fromSelected } from '../reducers';
import EditDashboardContent from './EditDashboardContent';
import DashboardVerticalOffset from './DashboardVerticalOffset';
import EditBar from '../ControlBarContainer/EditBar';
import {
    acSetEditNewDashboard,
    acSetEditDashboard,
} from '../actions/editDashboard';

class EditDashboard extends Component {
    state = {
        initialized: false,
    };

    initEditDashboard = () => {
        if (this.props.mode === 'new') {
            this.setState({ initialized: true });
            this.props.setNewDashboard();
        } else {
            if (this.props.dashboard) {
                this.setState({ initialized: true });
                this.props.setEditDashboard(
                    this.props.dashboard,
                    this.props.items
                );
            }
        }
    };

    componentDidMount() {
        this.initEditDashboard();
    }

    componentDidUpdate() {
        if (!this.state.initialized) {
            this.initEditDashboard();
        }
    }

    render() {
        return (
            <Fragment>
                <EditBar />
                <DashboardVerticalOffset />
                <EditDashboardContent />
            </Fragment>
        );
    }
}

const mapStateToProps = state => {
    const selectedId = fromSelected.sGetSelectedId(state);
    const dashboard = selectedId
        ? fromDashboards.sGetById(state, selectedId)
        : null;

    const items = fromDashboards.sGetItems(state);

    return {
        dashboard,
        items,
    };
};

export default connect(mapStateToProps, {
    setNewDashboard: acSetEditNewDashboard,
    setEditDashboard: acSetEditDashboard,
})(EditDashboard);
