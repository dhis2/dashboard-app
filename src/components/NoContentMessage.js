import PropTypes from 'prop-types'
import React from 'react'
import classes from './styles/NoContentMessage.module.css'

const NoContentMessage = ({ text }) => (
    <div className={classes.container}>{text}</div>
)

NoContentMessage.propTypes = {
    text: PropTypes.string,
}

export default NoContentMessage
