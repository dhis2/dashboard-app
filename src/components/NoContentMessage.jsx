import { colors } from '@dhis2/ui'
import PropTypes from 'prop-types'
import React from 'react'
import classes from './styles/NoContentMessage.module.css'

const NoContentMessage = ({ text }) => (
    <div className={classes.container}>
        <svg
            width="48"
            height="48"
            viewBox="0 0 48 48"
            fill="none"
            xmlns="http://www.w3.org/2000/svg"
        >
            <path
                d="M6.4387 47.8477L6.82684 45.8858C7.20466 45.9605 7.59673 46 8 46H10V48H8C7.46572 48 6.94367 47.9476 6.4387 47.8477ZM38 48V46H40C40.4033 46 40.7953 45.9605 41.1732 45.8858L41.5613 47.8477C41.0563 47.9476 40.5343 48 40 48H38ZM44.4449 46.6525L43.3323 44.9906C43.9875 44.552 44.552 43.9875 44.9906 43.3323L46.6525 44.4449C46.0685 45.3173 45.3173 46.0685 44.4449 46.6525ZM48 10H46V8C46 7.59673 45.9605 7.20466 45.8858 6.82684L47.8477 6.4387C47.9476 6.94367 48 7.46572 48 8V10ZM46.6525 3.55507L44.9906 4.66771C44.552 4.01254 43.9875 3.44804 43.3323 3.00941L44.4449 1.34748C45.3173 1.93154 46.0685 2.68266 46.6525 3.55507ZM10 0H8C7.46572 0 6.94367 0.0523755 6.4387 0.152272L6.82684 2.11425C7.20466 2.0395 7.59673 2 8 2H10V0ZM3.55508 1.34748L4.66771 3.00941C4.01254 3.44804 3.44804 4.01254 3.00941 4.66771L1.34748 3.55507C1.93154 2.68266 2.68266 1.93154 3.55508 1.34748ZM0 38H2V40C2 40.4033 2.0395 40.7953 2.11425 41.1732L0.152272 41.5613C0.0523755 41.0563 0 40.5343 0 40V38ZM3.55507 46.6525L4.66771 44.9906C4.01254 44.552 3.44804 43.9875 3.00941 43.3323L1.34748 44.4449C1.93154 45.3173 2.68266 46.0685 3.55507 46.6525ZM0 34H2V30H0V34ZM0 26H2V22H0V26ZM0 18H2V14H0V18ZM0 10H2V8C2 7.59673 2.0395 7.20466 2.11425 6.82684L0.152272 6.4387C0.0523756 6.94367 0 7.46572 0 8V10ZM14 0V2H18V0H14ZM22 0V2H26V0H22ZM30 0V2H34V0H30ZM38 0V2H40C40.4033 2 40.7953 2.0395 41.1732 2.11425L41.5613 0.152272C41.0563 0.0523755 40.5343 0 40 0H38ZM48 14H46V18H48V14ZM48 22H46V26H48V22ZM48 30H46V34H48V30ZM48 38H46V40C46 40.4033 45.9605 40.7953 45.8858 41.1732L47.8477 41.5613C47.9476 41.0563 48 40.5343 48 40V38ZM34 48V46H30V48H34ZM26 48V46H22V48H26ZM18 48V46H14V48H18Z"
                fill={colors.grey500}
            />
        </svg>

        <p>{text}</p>
    </div>
)

NoContentMessage.propTypes = {
    text: PropTypes.string,
}

export default NoContentMessage
