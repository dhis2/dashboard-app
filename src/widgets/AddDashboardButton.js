import React from 'react';
import AddIcon from '@material-ui/icons/AddCircle';
import { colors } from '../modules/colors';
import { NONAME } from 'dns';

const styles = {
    button: {
        width: 36,
        height: 36,
        padding: 0,
        backgroundColor: 'transparent',
        border: 'none',
        boxShadow: 'none',
        // cursor: 'pointer',
    },
    icon: {
        fill: colors.mediumGreen,
        // color: colors.mediumGreen,
        width: '1.3em',
        height: '1.3em',
        // cursor: 'pointer',
    },
};

const AddButton = () => {
    return (
        <button style={styles.button}>
            <AddIcon style={styles.icon} />
        </button>
    );
};

export default AddButton;
