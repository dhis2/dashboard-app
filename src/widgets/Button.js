import React from 'react';
import FlatButton from 'material-ui/FlatButton';
import { colors } from '../colors';

const style = {
    color: colors.royalBlue,
    backgroundColor: 'transparent',
    border: 'none',
    fontSize: '14px',
    fontWeight: 500,
    height: '36px',
    cursor: 'pointer',
};

const PlainButton = ({ onClick, children }) => {
    const label = typeof children === 'string' ? children : null;

    return (
        <FlatButton
            style={style}
            onClick={onClick}
            label={label}
            disabledBackgroundColor={colors.white}
            disabledLabelColor={colors.lightMediumGrey}
        />
    );
};

export default PlainButton;
