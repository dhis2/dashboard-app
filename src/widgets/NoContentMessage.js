import React from 'react';
import { colors } from '../colors';

export const NoContentMessage = ({ text }) => (
    <div
        style={{
            padding: '50px 10px',
            textAlign: 'center',
            fontSize: '15px',
            fontWeight: 500,
            color: colors.charcoalGrey,
        }}
    >
        {text}
    </div>
);

export default NoContentMessage;
