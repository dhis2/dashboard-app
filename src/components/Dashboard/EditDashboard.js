import React, { Component } from 'react'
import { connect } from 'react-redux'
import i18n from '@dhis2/d2-i18n'
import PropTypes from 'prop-types'

import { acSetEditDashboard } from '../../actions/editDashboard'
import { sGetSelectedId } from '../../reducers/selected'
import {
    sGetDashboardById,
    sGetDashboardItems,
    sDashboardsIsFetching,
} from '../../reducers/dashboards'
import DashboardVerticalOffset from './DashboardVerticalOffset'
import DashboardContent from './DashboardContent'
import EditBar from '../ControlBar/EditBar'
import NoContentMessage from '../../widgets/NoContentMessage'

export const Content = ({ updateAccess }) => {
    return updateAccess ? (
        <DashboardContent editMode={true} />
    ) : (
        <NoContentMessage text={i18n.t('No access')} />
    )
}

Content.propTypes = {
    updateAccess: PropTypes.bool,
}
export class EditDashboard extends Component {
    state = {
        initialized: false,
    }

    initEditDashboard = () => {
        if (this.props.dashboard) {
            this.setState({ initialized: true })
            this.props.setEditDashboard(this.props.dashboard, this.props.items)
        }
    }

    componentDidMount() {
        this.initEditDashboard()
    }

    componentDidUpdate() {
        if (!this.state.initialized) {
            this.initEditDashboard()
        }
    }

    getDashboardContent = () => {
        const contentNotReady =
            !this.props.dashboardsLoaded || this.props.id === null

        return (
            <div className="dashboard-wrapper">
                {contentNotReady ? null : (
                    <Content updateAccess={this.props.updateAccess} />
                )}
            </div>
        )
    }

    render() {
        return (
            <>
                <EditBar />
                <DashboardVerticalOffset editMode={true} />
                {this.getDashboardContent()}
            </>
        )
    }
}

EditDashboard.propTypes = {
    dashboard: PropTypes.object,
    dashboardsLoaded: PropTypes.bool,
    id: PropTypes.string,
    items: PropTypes.array,
    setEditDashboard: PropTypes.func,
    updateAccess: PropTypes.bool,
}

const mapStateToProps = state => {
    const id = sGetSelectedId(state)
    const dashboard = id ? sGetDashboardById(state, id) : null

    const updateAccess =
        dashboard && dashboard.access ? dashboard.access.update : false

    return {
        dashboard,
        id,
        updateAccess,
        items: sGetDashboardItems(state),
        dashboardsLoaded: !sDashboardsIsFetching(state),
    }
}

export default connect(mapStateToProps, {
    setEditDashboard: acSetEditDashboard,
})(EditDashboard)
