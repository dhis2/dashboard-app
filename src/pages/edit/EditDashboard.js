import { useDataEngine } from '@dhis2/app-runtime'
import i18n from '@dhis2/d2-i18n'
import { Layer, CenteredContent, CircularLoader } from '@dhis2/ui'
import cx from 'classnames'
import PropTypes from 'prop-types'
import React, { useState, useEffect } from 'react'
import { connect } from 'react-redux'
import { Redirect } from 'react-router-dom'
import { acSetEditDashboard } from '../../actions/editDashboard'
import { apiFetchDashboard } from '../../api/fetchDashboard'
import DashboardContainer from '../../components/DashboardContainer'
import NoContentMessage from '../../components/NoContentMessage'
import { useWindowDimensions } from '../../components/WindowDimensionsProvider'
import { EDIT } from '../../modules/dashboardModes'
import { setHeaderbarVisible } from '../../modules/setHeaderbarVisible'
import { isSmallScreen } from '../../modules/smallScreen'
import { sGetIsPrintPreviewView } from '../../reducers/editDashboard'
import LayoutPrintPreview from '../print/PrintLayoutDashboard'
import ActionsBar from './ActionsBar'
import ItemGrid from './ItemGrid'
import NotSupportedNotice from './NotSupportedNotice'
import classes from './styles/EditDashboard.module.css'
import TitleBar from './TitleBar'

const EditDashboard = props => {
    const dataEngine = useDataEngine()
    const { width } = useWindowDimensions()
    const [redirectUrl, setRedirectUrl] = useState(null)
    const [hasUpdateAccess, setHasUpdateAccess] = useState(null)
    const [isLoading, setIsLoading] = useState(true)

    useEffect(() => {
        const loadDashboard = async () => {
            try {
                const dashboard = await apiFetchDashboard(
                    dataEngine,
                    props.id,
                    { mode: EDIT }
                )
                props.setEditDashboard(dashboard)
                setHasUpdateAccess(dashboard.access?.update || false)
                setIsLoading(false)
            } catch (error) {
                setRedirectUrl(props.id ? `/${props.id}` : '/')
                setIsLoading(false)
            }
        }

        if (isSmallScreen(width)) {
            setRedirectUrl(props.id ? `/${props.id}` : '/')
            return
        }
        setHeaderbarVisible(true)

        loadDashboard()
    }, [props.id])

    if (redirectUrl) {
        return <Redirect to={redirectUrl} />
    }

    if (isLoading) {
        return (
            <Layer translucent>
                <CenteredContent>
                    <CircularLoader />
                </CenteredContent>
            </Layer>
        )
    }

    const renderGrid = () => {
        if (props.isPrintPreviewView) {
            return <LayoutPrintPreview fromEdit={true} />
        }
        return (
            <DashboardContainer>
                <TitleBar />
                <ItemGrid />
            </DashboardContainer>
        )
    }

    return (
        <>
            <div
                className={cx(classes.container, 'dashboard-scroll-container')}
                data-test="outer-scroll-container"
            >
                <ActionsBar />
                {hasUpdateAccess ? (
                    renderGrid()
                ) : (
                    <NoContentMessage text={i18n.t('No access')} />
                )}
            </div>
            <div className={classes.notice}>
                <NotSupportedNotice
                    message={i18n.t(
                        'Editing dashboards on small screens is not supported. Resize your screen to return to edit mode.'
                    )}
                />
            </div>
        </>
    )
}

EditDashboard.propTypes = {
    id: PropTypes.string,
    isPrintPreviewView: PropTypes.bool,
    setEditDashboard: PropTypes.func,
}

const mapStateToProps = (state, ownProps) => ({
    id: ownProps.match?.params?.dashboardId || null,
    isPrintPreviewView: sGetIsPrintPreviewView(state),
})

export default connect(mapStateToProps, {
    setEditDashboard: acSetEditDashboard,
})(EditDashboard)
