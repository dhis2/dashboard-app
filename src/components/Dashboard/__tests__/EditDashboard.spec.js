import React from 'react';
import { shallow } from 'enzyme';

import { EditDashboard, Content } from '../EditDashboard';
import { NoContentMessage } from '../../../widgets/NoContentMessage';

jest.mock('../DashboardContent', () => () => <div />);
jest.mock('../../ControlBar/EditBar', () => () => <div />);
jest.mock('../DashboardVerticalOffset', () => () => <div />);

describe('EditDashboard', () => {
    let props;
    let shallowEditDashboard;
    const editDashboard = () => {
        if (!shallowEditDashboard) {
            shallowEditDashboard = shallow(<EditDashboard {...props} />);
        }
        return shallowEditDashboard;
    };

    const assertContent = hasContent => {
        const content = editDashboard().find(Content);

        expect(content.length).toBe(1);
        expect(content.dive().find(NoContentMessage)).toHaveLength(
            hasContent ? 0 : 1
        );
    };

    beforeEach(() => {
        props = {
            dashboard: undefined,
            id: undefined,
            updateAccess: undefined,
            items: undefined,
            dashboardsLoaded: undefined,
        };
        shallowEditDashboard = undefined;
    });

    describe('when "dashboardsLoaded" is false', () => {
        it('does not render any children inside the div', () => {
            props.dashboardsLoaded = false;

            expect(
                editDashboard()
                    .find('.dashboard-wrapper')
                    .children().length
            ).toBe(0);
        });
    });

    describe('when "dashboardsLoaded" is true', () => {
        beforeEach(() => {
            props.dashboardsLoaded = true;
        });

        describe('when "id" is null', () => {
            it('does not render any children inside the div', () => {
                props.id = null;
                expect(
                    editDashboard()
                        .find('.dashboard-wrapper')
                        .children().length
                ).toBe(0);
            });
        });

        describe('when id is not null', () => {
            beforeEach(() => {
                props.id = 'abc123';
            });

            describe('when updateAccess is true', () => {
                it('renders DashboardContent', () => {
                    props.updateAccess = true;
                    assertContent(true);
                });
            });

            describe('when updateAccess is false', () => {
                it('renders a NoContentMessage', () => {
                    props.updateAccess = false;
                    assertContent(false);
                });
            });
        });
    });
});
