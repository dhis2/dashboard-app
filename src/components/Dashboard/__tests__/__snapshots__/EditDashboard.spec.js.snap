// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditDashboard renders dashboard 1`] = `
<div>
  <div
    class="container"
  >
    <div>
      EditBar
    </div>
    <div
      class="dashboard-wrapper"
    >
      <div>
        EditTitleBar
      </div>
      <div>
        EditItemGrid
      </div>
    </div>
  </div>
  <div
    class="notice"
  >
    <div
      class="jsx-761413583 centered-content top"
      data-test="dhis2-uicore-centeredcontent"
    >
      <div
        class="jsx-761413583 centered-inner-content"
      >
        <div
          class="noticeContainer"
        >
          <div
            class="jsx-3561221191 root warning"
            data-test="dhis2-uicore-noticebox"
          >
            <div
              data-test="dhis2-uicore-noticebox-icon"
            >
              <svg
                class="jsx-884574500 jsx-1769496071"
                viewBox="0 0 48 48"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  class="jsx-884574500"
                  d="M2 42h44L24 4 2 42zm24-6h-4v-4h4v4zm0-8h-4v-8h4v8z"
                />
              </svg>
            </div>
            <div
              class="jsx-3561221191"
            >
              <h6
                class="jsx-1234954849"
                data-test="dhis2-uicore-noticebox-title"
              >
                Not supported
              </h6>
              <div
                class="jsx-3638294246"
                data-test="dhis2-uicore-noticebox-message"
              >
                Editing dashboards on small screens is not supported. Resize your screen to return to edit mode.
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`EditDashboard renders message when not enough access 1`] = `
<div>
  <div
    class="container"
  >
    <div>
      EditBar
    </div>
    <div
      style="padding: 50px 10px; text-align: center; font-size: 15px; font-weight: 500; color: rgb(110, 122, 138);"
    >
      No access
    </div>
  </div>
  <div
    class="notice"
  >
    <div
      class="jsx-761413583 centered-content top"
      data-test="dhis2-uicore-centeredcontent"
    >
      <div
        class="jsx-761413583 centered-inner-content"
      >
        <div
          class="noticeContainer"
        >
          <div
            class="jsx-3561221191 root warning"
            data-test="dhis2-uicore-noticebox"
          >
            <div
              data-test="dhis2-uicore-noticebox-icon"
            >
              <svg
                class="jsx-884574500 jsx-1769496071"
                viewBox="0 0 48 48"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  class="jsx-884574500"
                  d="M2 42h44L24 4 2 42zm24-6h-4v-4h4v4zm0-8h-4v-8h4v8z"
                />
              </svg>
            </div>
            <div
              class="jsx-3561221191"
            >
              <h6
                class="jsx-1234954849"
                data-test="dhis2-uicore-noticebox-title"
              >
                Not supported
              </h6>
              <div
                class="jsx-3638294246"
                data-test="dhis2-uicore-noticebox-message"
              >
                Editing dashboards on small screens is not supported. Resize your screen to return to edit mode.
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`EditDashboard renders print preview 1`] = `
<div>
  <div
    class="container"
  >
    <div>
      EditBar
    </div>
    <div>
      LayoutPrintPreview
    </div>
  </div>
  <div
    class="notice"
  >
    <div
      class="jsx-761413583 centered-content top"
      data-test="dhis2-uicore-centeredcontent"
    >
      <div
        class="jsx-761413583 centered-inner-content"
      >
        <div
          class="noticeContainer"
        >
          <div
            class="jsx-3561221191 root warning"
            data-test="dhis2-uicore-noticebox"
          >
            <div
              data-test="dhis2-uicore-noticebox-icon"
            >
              <svg
                class="jsx-884574500 jsx-1769496071"
                viewBox="0 0 48 48"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  class="jsx-884574500"
                  d="M2 42h44L24 4 2 42zm24-6h-4v-4h4v4zm0-8h-4v-8h4v8z"
                />
              </svg>
            </div>
            <div
              class="jsx-3561221191"
            >
              <h6
                class="jsx-1234954849"
                data-test="dhis2-uicore-noticebox-title"
              >
                Not supported
              </h6>
              <div
                class="jsx-3638294246"
                data-test="dhis2-uicore-noticebox-message"
              >
                Editing dashboards on small screens is not supported. Resize your screen to return to edit mode.
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
